doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The required changes necessary to NumPy are massive and touch a large part
+of the code base.
+We thus propose the implementation of the above steps listed in Phases I and II.
+
+Although it is possible that new DTypes will only be useful after Phase II
+is finished, this NEP proposes to start implementation in an incremental way.
+This means that in a first step the ``DType`` classes will be added, with
+all, or most, new exposed API points giving a ``PreliminaryDTypeAPIWarning``.
+
+This allows for smaller patches and further future changes. In these first
+steps, only a very limited new C-API will be exposed to allow writing tests,
+with no expectation that early adopters will begin using it.
+The addition of ``DType`` will then enable addressing other changes
+more incrementally:
+
+1. New machinery for array coercion, with the goal of enabling user DTypes
+   with appropriate class methods.
+2. The replacement or wrapping of the current casting machinery.
+3. Incremental redefinition of the current ``ArrFunctions`` slots into
+   DType method slots.
+
+Parallel to these, after step 1. is finished, Phase II of revising the
+UFunc machinery can be started.
+
+In particular the step of creating a C defined ``DTypeMeta`` class with its
+instances being ``DTypeClasses`` is a necessary first step to add the ability
+of storing custom slots on the datatype in C.
+This ``DTypeMeta`` must thus be implemented before being widely used to
+restructure or enhance current code, thus we propose to proceed with mainly
+private additions to the DType classes.
+
+
+Backward compatibility
+----------------------
+
+While the actual backward compatibility impact is not yet fully clear,
+we anticipate, and accept the following changes:
+
+* **Python API**:
+
+  * ``type(np.dtype("f8"))`` will be a subclass of ``np.dtype``, while right
+    now ``type(np.dtype("f8")) is np.dtype``.
+    Code should use ``isinstance`` checks, and in very rare cases may have to
+    be adapted to use it.
+
+* **C-API**:
+
+    * In old versions of NumPy ``PyArray_DescrCheck`` is a macro which uses
+      ``type(dtype) is np.dtype``. When compiling against an old NumPy version,
+      the macro may have to be replaced with the corresponding
+      ``PyObject_IsInstance`` call. (If this is a problem, we could backport
+      fixing the macro)
+
+   * The UFunc machinery changes will break *limited* parts of the current
+     implementation. Replacing e.g. the default ``TypeResolver`` is expected
+     to remain supported for a time, although optimized masked inner loop iteration
+     (which is not even used *within* NumPy) is expected to not remain supported
+     and lead to errors instead.
+
+   * All current functions defined on the dtypes currently, such as
+     ``PyArray_Descr->f->nonzero``, will move to a new home and recieve new
+     ways to access them.
+     This means that in the long run some (rare) lowlevel access code will
```

Comment:
Consider removing "(rare)" and adding a new sentence at the end: "Such changes are expected to be necessary very infrequently"